### PR TITLE
*Non-Boussinesq revision of energetic_PBL

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
This commit revises the internal routines called by energetic_PBL to work in fully non-Boussinesq mode, eliminating all dependencies on the Boussinesq reference density when in non-Boussinesq mode.  The publicly visible interfaces to this module and the external routines it calls have already been revised, so only this file needs to be updated.  The specific changes include:

 - Work with both thickness (h) and height change (dz) variables in energetic_PBL using thickness_to_dz to translate between the two.  In Boussinesq mode, these are a simple rescaling by a constant factor, but in non-Boussinesq mode they vary by a factor of the specific volume.  Some calculations are unnecessarily duplicated in Boussinesq mode, so this might slow the model slightly, but they are not duplicative in non-Boussinesq mode.

 - When in non-Boussinesq mode, use forces%tau_mag and tv%SpV_avg instead of forces%ustar and GV%Rho0 to determine the friction velocity and mechanical TKE input used in energetic_PBL, and to convert the local TKE to a turbulent velocity.  This includes the addition of a new energy input argument and an interface specific volume argument to ePBL_column and the removal of an unused argument to find_mstar.

 - Use tau_mag_gustless times the in situ specfic volume instead of ustar_gustless to calculate the Langmuir number used by ePBL when it is run in non-Boussinesq mode.

 - The unused Ustar_mean argument to find_mstar was removed.

  This change involves the addition of 2 new arguments to ePBL_column and changes to the units of another argument.  There are changes to the units of 2 elements of the energetic_PBL_CS type and 14 internal variables. This change includes the addition of 13 new internal variables, the removal of 8 internal variable (most of which were renamed to reflect their new units). With this change, 9 thickness to height conversion factors have been eliminated, and GV%Rho0 is only used in Boussinesq mode.

  All answers in Boussinesq mode are bitwise identical, but they change in non-Boussinesq mode cases that use ePBL, which no longer depend on the value of the Boussinesq reference density.

### CODE REVIEW

File: src/parameterizations/vertical/MOM_energetic_PBL.F90 [hunk 1]
```diff
@@ -619,6 +669,8 @@ subroutine ePBL_column(h, dz, u, v, T0, S0, dSV_dT, dSV_dS, TKE_forcing, B_flux,
     MixLen_shape, & ! A nondimensional shape factor for the mixing length that
                     ! gives it an appropriate asymptotic value at the bottom of
                     ! the boundary layer [nondim].
+    h_dz_int, &     ! The ratio of the layer thicknesses over the vertical distances
+                    ! across the layers surrounding an interface [H Z-1 ~> nonodim or kg m-3]
```
Comment (breichl) [inline src/parameterizations/vertical/MOM_energetic_PBL.F90:673]: Note a typo: nonodim

File: src/parameterizations/vertical/MOM_energetic_PBL.F90 [hunk 2]
```diff
@@ -531,12 +576,16 @@ subroutine ePBL_column(h, dz, u, v, T0, S0, dSV_dT, dSV_dS, TKE_forcing, B_flux,
...
+                                                   !! step before the application of the efficiency
+                                                   !! in mstar. [R Z3 T-2 ~> J m-2].
   real,                    intent(inout) :: MLD_io !< A first guess at the mixed layer depth on input, and
-                                                   !! the calculated mixed layer depth on output [Z ~> m].
+                                                   !! the calculated mixed layer depth on output [Z ~> m]
   real,                    intent(in)    :: dt     !< Time increment [T ~> s].
   real, dimension(SZK_(GV)+1), &
                            intent(out)   :: Kd     !< The diagnosed diffusivities at interfaces
-                                                   !! [Z2 T-1 ~> m2 s-1].
+                                                   !! [H Z T-1 ~> kg m-1 s-1].
```
Comment (breichl) [inline src/parameterizations/vertical/MOM_energetic_PBL.F90:588]: Units should be [H Z T-1 ~> m2 s-1 or kg m-1 s-1].

**Comment by breichl** [comment]
@Hallberg-NOAA I noted a typo and a missing Boussinesq "or" non-Boussinesq unit missing in the comments, but the code changes look to be fine.

**Comment by Hallberg-NOAA** [comment]
Thank you for spotting those inconsistent comments.  I have now corrected them as suggested in an updated version of this PR that also rebases it on top of the latest version of dev/gfdl.

**Review by marshallward — APPROVED** [review]

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/20726 :heavy_check_mark:

**Pull request merged**